### PR TITLE
Add port exposure RPC support

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add ziti client support for configs, service policies, and device identities
- update server handlers/converters for new port exposure RPCs
- expand client/converter tests for new flows

## Testing
- go vet ./...
- go test ./...

Refs #37

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- go vet ./...
- go test ./...

Tests: passed 2 packages, failed 0, skipped 0.
Lint: no issues.

**Comment by rowan-stein** [comment]
Requesting review for this PR. Implements the port exposure additions per issue #37:
- 5 new RPCs: CreateServicePolicy, DeleteServicePolicy, DeleteService, CreateDeviceIdentity, DeleteDeviceIdentity
- Updated CreateService to support host.v1/intercept.v1 configs
- Refactored App/Runner identity lifecycle to decouple service creation
- New tests for all additions

CI is pending.

File: internal/ziti/client.go [hunk 1]
```diff
@@ -100,7 +188,51 @@ func (c *Client) CreateAgentIdentity(ctx context.Context, agentID uuid.UUID) (st
...
+	}
+
+	jwt, err := c.fetchEnrollmentJWT(ctx, zitiID)
+	if err != nil {
+		return "", "", err
+	}
+	return zitiID, jwt, nil
+}
+
+func (c *Client) CreateDeviceIdentity(ctx context.Context, userIdentityID uuid.UUID, name string) (string, string, error) {
```
Comment (noa-lucent) [inline internal/ziti/client.go:286]: [major] **Duplicate identity-creation + response-extraction blocks.** `CreateAgentIdentity`, `CreateDeviceIdentity`, `CreateAndEnrollAppIdentity`, `CreateAndEnrollRunnerIdentity`, and `CreateAndEnrollServiceIdentity` all repeat:

1. Build `IdentityCreate`, call `CreateIdentity` via `withReauth`, unwrap `Payload.Data.ID` with nil checks.
2. Then either `fetchEnrollmentJWT` or `enrollIdentity`.

There are **8 copies** of the `created.Payload == nil || created.Payload.Data == nil` block (6 new in this PR) and nearly identical create-then-check-response sequences.

Extract a shared helper, e.g.:

```go
func (c *Client) createIdentity(ctx context.Context, params *identity.CreateIdentityParams) (string, error) {
    var created *identity.CreateIdentityCreated
    err := c.withReauth(func() error {
        var callErr error
        created, callErr = c.identity.CreateIdentity(params, nil)
        return callErr
    })
    if err != nil {
        return "", fmt.Errorf("create ziti identity: %w", err)
    }
    if created.Payload == nil || created.Payload.Data == nil {
        return "", errors.New("create ziti identity response missing data")
    }
    if created.Payload.Data.ID == "" {
        return "", errors.New("create ziti identity response missing id")
    }
    return created.Payload.Data.ID, nil
}
```

Then each public method only assembles `IdentityCreate` params and calls `createIdentity`. Same idea applies to the config/service/service-policy creation patterns. This reduces risk of drift and makes each method's unique logic visible.

File: internal/server/server.go [hunk 2]
```diff
@@ -57,6 +60,150 @@ func (s *Server) CreateAgentIdentity(ctx context.Context, req *zitimanagementv1.
...
+	if err := s.ziti.DeleteService(ctx, zitiServiceID); err != nil {
+		if errors.Is(err, ziti.ErrServiceNotFound) {
+			log.Printf("ziti service %s already deleted", zitiServiceID)
+		} else {
+			return nil, status.Errorf(codes.Internal, "delete ziti service: %v", err)
+		}
+	}
+
+	return &zitimanagementv1.DeleteRunnerIdentityResponse{}, nil
+}
```
Comment (noa-lucent) [inline internal/server/server.go:184]: [major] **Near-identical `DeleteRunnerIdentity` and `DeleteAppIdentity` handlers.** The logic body is the same: parse `identity_id`, validate `ziti_service_id`, resolve from store, delete managed identity, delete ziti identity (idempotent on not-found), delete ziti service (idempotent on not-found).

The only difference is the return type (`DeleteRunnerIdentityResponse` vs `DeleteAppIdentityResponse`). Extract the shared teardown into a helper:

```go
func (s *Server) deleteManagedIdentityAndService(ctx context.Context, identityID uuid.UUID, zitiServiceID string) error {
    identity, err := s.store.ResolveIdentityByIdentityID(ctx, identityID)
    if err != nil {
        return toStatusError(err)
    }
    if err := s.store.DeleteManagedIdentity(ctx, identity.ZitiIdentityID); err != nil {
        return toStatusError(err)
    }
    if err := s.ziti.DeleteIdentity(ctx, identity.ZitiIdentityID); err != nil {
        if errors.Is(err, ziti.ErrIdentityNotFound) {
            log.Printf("ziti identity %s already deleted", identity.ZitiIdentityID)
        } else {
            return status.Errorf(codes.Internal, "delete ziti identity: %v", err)
        }
    }
    if err := s.ziti.DeleteService(ctx, zitiServiceID); err != nil {
        if errors.Is(err, ziti.ErrServiceNotFound) {
            log.Printf("ziti service %s already deleted", zitiServiceID)
        } else {
            return status.Errorf(codes.Internal, "delete ziti service: %v", err)
        }
    }
    return nil
}
```

Each RPC handler then just validates input and delegates. This avoids duplication and reduces the risk of the two handlers drifting apart.

File: internal/ziti/client.go [hunk 3]
```diff
@@ -100,7 +188,51 @@ func (c *Client) CreateAgentIdentity(ctx context.Context, agentID uuid.UUID) (st
...
+	identityType := rest_model.IdentityTypeDevice
+	isAdmin := false
+	roleAttrs := rest_model.Attributes{"devices"}
+	externalID := userIdentityID.String()
+	params := identity.NewCreateIdentityParamsWithContext(ctx)
```
Comment (noa-lucent) [inline internal/ziti/client.go]: [minor] **Magic string `"devices"` for role attribute.** The role attribute is a policy decision that should be a named constant (alongside `"agents"`, `"apps"`, etc.). This makes the value searchable and documents intent:

```go
const roleDevices = "devices"
```

The same applies to the other role-attribute strings scattered throughout (e.g., `"agents"`, `"apps"` on lines 184, 498). Consider centralizing these as package-level constants.

File: internal/server/converter.go [hunk 4]
```diff
@@ -42,27 +45,117 @@ func fromProtoServiceType(value zitimanagementv1.ServiceType) (store.ServiceType
...
+	}
+	port := value.GetPort()
+	if port <= 0 {
+		return nil, fmt.Errorf("port must be positive")
+	}
+	return &ziti.HostV1ConfigData{
+		Protocol: protocol,
+		Address:  address,
+		Port:     port,
+	}, nil
```
Comment (noa-lucent) [inline internal/server/converter.go:92]: [nit] `port <= 0` rejects port 0 but also doesn't upper-bound the value. Since `Port` is `int32`, values > 65535 would pass validation here and be sent to OpenZiti. Consider adding `port > 65535` to the check. Same applies to port ranges in `fromProtoInterceptV1Config` (line 126).

File: internal/server/converter_test.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+			}
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if got != tc.want {
+				t.Fatalf("expected %q, got %q", tc.want, got)
+			}
+		})
+	}
+}
```
Comment (noa-lucent) [inline internal/server/converter_test.go:121]: [minor] **Missing converter tests for `fromProtoHostV1Config` and `fromProtoInterceptV1Config`.** These are non-trivial boundary converters with multiple validation branches (empty protocol, empty address, port ≤ 0, empty protocols list, empty addresses list, empty port ranges, high < low, etc.). The issue spec explicitly calls for converter tests, and these two functions are the most complex converters added. Please add table-driven tests covering at least the valid case and each error branch.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Overall the implementation correctly covers the issue requirements — new RPCs, updated signatures, config/service-policy support, and tests for the new flows. The architecture is clean: boundary validation in server handlers, internal logic in the ziti client.

Two major concerns before merging:

1. **Significant code duplication in `client.go`** — the identity-creation + response-extraction pattern is repeated 5+ times (8 copies of the nil-check block, 6 new in this PR). Extract shared helpers to reduce drift risk.

2. **Near-identical `DeleteAppIdentity` / `DeleteRunnerIdentity` handlers** — same logic, different return types. Extract the shared teardown.

Also requesting converter test coverage for the new `fromProtoHostV1Config` and `fromProtoInterceptV1Config` functions, and minor items around magic strings and port validation bounds.

**Comment by casey-brooks** [comment]
## Summary
- reconciled ziti client helpers to match main's client structure while keeping create/config/service policy refactors and role-attribute constants in `internal/ziti/client.go`.

## Testing
- go vet ./...
- go test ./...

Tests: 2 passed, 0 failed, 0 skipped.
Lint: go vet ./... (no issues).

**Comment by rowan-stein** [comment]
All review feedback addressed:
1. Extracted shared `createIdentity` helper + config/service/service-policy creation helpers in `client.go`
2. Extracted `deleteManagedIdentityAndService` helper in `server.go`
3. Role-attribute magic strings replaced with named constants
4. Added table-driven converter tests for `fromProtoHostV1Config` and `fromProtoInterceptV1Config`
5. Added port upper-bound validation (65535)

CI passed. Ready for re-review.

**Review by noa-lucent — APPROVED** [review]
All five prior review comments have been addressed:

1. ✅ **Duplicate identity-creation blocks** — extracted `createWithReauth`, `extractCreateID`, `createIdentity`, and `createAndEnrollIdentity` helpers. Zero remaining copies of the inline nil-check block.
2. ✅ **Duplicate delete handlers** — extracted `deleteManagedIdentityAndService` helper, both `DeleteAppIdentity` and `DeleteRunnerIdentity` now delegate to it.
3. ✅ **Magic role-attribute strings** — centralized as `roleAttributeAgents`, `roleAttributeApps`, `roleAttributeDevices` constants.
4. ✅ **Port validation bounds** — added `maxPort = 65535` constant and upper-bound checks for both `fromProtoHostV1Config` and `fromProtoInterceptV1Config`.
5. ✅ **Missing converter tests** — comprehensive table-driven tests added for `fromProtoHostV1Config` and `fromProtoInterceptV1Config`, covering nil, valid, and all error branches.

The mutex-guarded accessor pattern for reauthentication is a nice addition for thread safety. Code is clean and well-structured. LGTM.
